libs/gltfio/src/PathTracer.cpp

```diff
@@ -0,0 +1,282 @@
+    return *this;
+
+}
+
+PathTracer::Builder& PathTracer::Builder::uvCamera(const char* uvAttribute) {
+    mPathTracer->mUvCamera = uvAttribute;
+    return *this;
+
+}
+
+PathTracer::Builder& PathTracer::Builder::tileCallback(TileCallback onTile, void* userData) {
+    mPathTracer->mTileCallback = onTile;
+    mPathTracer->mTileUserData = userData;
+    return *this;
+}
+
+PathTracer::Builder& PathTracer::Builder::doneCallback(DoneCallback onDone, void* userData) {
+    mPathTracer->mDoneCallback = onDone;
+    mPathTracer->mDoneUserData = userData;
+    return *this;
+}
+
+PathTracer PathTracer::Builder::build() {
+    return *mPathTracer;
+}
+
+#ifndef FILAMENT_HAS_EMBREE
+
+bool PathTracer::render() {
+    puts("Embree is not available.");
+    return false;
+}
+
+#else
+
+// Note that the embree context can outlive the user-facing PathTracer object due to work
+// occuring on multiple background threads.
+struct EmbreeContext {
+    image::LinearImage renderTarget;
+    const cgltf_data* sourceAsset;
+    SimpleCamera filmCamera;
+    const char* uvCamera;
+    PathTracer::TileCallback tileCallback;
+    void* tileUserData;
+    PathTracer::DoneCallback doneCallback;
+    void* doneUserData;
+    std::atomic<int> numRemainingTiles;
+    RTCDevice embreeDevice;
+    RTCScene embreeScene;
+};
+
+static void renderTile(EmbreeContext* context, PixelRectangle rect) {
+    image::LinearImage& image = context->renderTarget;
+
+    // Precompute some camera parameters.
+    const SimpleCamera& camera = context->filmCamera;
+    const float iw = 1.0f / image.getWidth();
+    const float ih = 1.0f / image.getHeight();
+    const float theta = camera.vfovDegrees * M_PI / 180;
+    const float f = tanf(theta / 2);
+    const float a = camera.aspectRatio;
+    const float3 org = camera.eyePosition;
+    const uint16_t hm1 = image.getHeight() - 1;
+
+    // Compute the camera basis: view, right, and up vectors.
+    const float3 v = normalize(camera.targetPosition - org);
+    const float3 r = normalize(cross(v, camera.upVector));
+    const float3 u = cross(r, v);
+
+    // Given a pixel row and column, generate a ray from the eye through the film.
+    auto generateCameraRay = [=] (uint16_t row, uint16_t col) {
+        const uint16_t x = col;
+        const uint16_t y = hm1 - row;
+        const float s = (2.0f * (x + 0.5f) * iw - 1.0f);
+        const float t = (2.0f * (y + 0.5f) * ih - 1.0f);
+        const float3 dir = normalize(a * f * s * r - f * t * u + v);
+        return RTCRay {
+            .org_x = org.x,
+            .org_y = org.y,
+            .org_z = org.z,
+            .tnear = 0,
+            .dir_x = dir.x,
+            .dir_y = dir.y,
+            .dir_z = dir.z,
+            .time = 0,
+            .tfar = inf,
+            .mask = 0xffffffff
+        };
+    };
+
+    // Loop over all pixels in the tile.
+    for (uint16_t row = rect.topLeft.y, len = rect.bottomRight.y; row < len; ++row) {
```

Comment:
it's generally faster to use size_t for loop counters -- compiler doesn't have to mask. It's possible it'll do the right thing, but I've seen case where it didn't.